### PR TITLE
Add recipe for HoloPy

### PR DESCRIPTION
**Pull request by barkls**
This is a follow-up to #2029, now that we have completed a release of our package. 

We have temporarily given up on compiling Fortran on Windows systems. If there has been any recent progress on this front, please let me know and I will re-tackle the issue.

### CODE REVIEW

File: recipes/holopy/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  fn: {{ name|lower }}-{{ version }}.tar.gz
+  url: https://github.com/manoharan-lab/{{ name }}/archive/{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  skip: True  # [py2k or win or np<112]
+
+requirements:
+  build:
```
Comment (jakirkham) [inline recipes/holopy/meta.yaml]: Please add `toolchain`.

File: recipes/holopy/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    - python
```
Comment (jakirkham) [inline recipes/holopy/meta.yaml]: Please add `toolchain`.

File: recipes/holopy/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    - numpy x.x
+    - setuptools
+    - gcc
```
Comment (jakirkham) [inline recipes/holopy/meta.yaml]: Why is `gcc` required?

File: recipes/holopy/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  version: {{ version }}
+
+source:
+  fn: {{ name|lower }}-{{ version }}.tar.gz
+  url: https://github.com/manoharan-lab/{{ name }}/archive/{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+build:
+  number: 0
+  skip: True  # [py2k or win or np<112]
```
Comment (jakirkham) [inline recipes/holopy/meta.yaml]: Could you please elaborate on these constraints?

File: recipes/holopy/build.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+if [ $(uname) == Darwin ]; then
+    export CFLAGS="-arch x86_64"
+    export FFLAGS="-static -ff2c -arch x86_64"
+    export LDFLAGS="-Wall -undefined dynamic_lookup -bundle -arch x86_64"
+    export LDFLAGS="-Wl,-search_paths_first -L$(pwd) $LDFLAGS"
```
Comment (jakirkham) [inline recipes/holopy/build.sh]: Please make sure the existing environment variables are preserved (e.g. `export CFLAGS="${CFLAGS} -arch x86_64"`).

Comment (jakirkham) [inline recipes/holopy/build.sh]: Also could you please elaborate on the need for these flags?

File: recipes/holopy/meta.yaml (hunk 2 above)
Comment (barkls) [inline recipes/holopy/meta.yaml]: When I add toolchain as a build requirement, f2py compiles Fortran code properly, but throws the following error when trying to compile C:

``In function `_start': (.text+0x20): undefined reference to `main'``

@jakirkham any insight on this? What is toolchain supposed to do?

File: recipes/holopy/meta.yaml (hunk 3 above)
Comment (barkls) [inline recipes/holopy/meta.yaml]: We need gcc to compile Fortran code.

File: recipes/holopy/meta.yaml (hunk 4 above)
Comment (barkls) [inline recipes/holopy/meta.yaml]: Our package only runs with python 3.

We cannot figure out how to compile Fortran on Windows.

Something happened when xarray upgraded to 0.9 that interferes with how we call out to our Fortran code for numpy 1.11. The simplest fix was to require users to have numpy >=1.12.

File: recipes/holopy/build.sh (hunk 5 above)
Comment (barkls) [inline recipes/holopy/build.sh]: I copied this file from another feedstock compiling Fortran code (I can't remember which one). I believe it has something to do with specifying where to find gfortran libraries. 

The code doesn't compile properly if I don't include it.

File: recipes/holopy/meta.yaml (hunk 2 above)
Comment (jakirkham) [inline recipes/holopy/meta.yaml]: Unfortunately that is not enough to go on. Do you have a log? Maybe try pushing this change so we can see the failure on Travis CI.

**Comment by jakirkham** [comment]
Toggling for CI.

**Comment by jakirkham** [comment]
Could you please disable CircleCI on the `manoharan-lab` fork of `staged-recipes`, @barkls?

**Comment by barkls** [comment]
any movement here @jakirkham? The builds passed without toolchain, and I'm not sure why they are failing with it. Any help towards getting this feedstock up is appreciated.

**Comment by jakirkham** [comment]
There is one test failure. I'm not convinced the `toolchain` is the cause of the issue, but I wouldn't be surprised if the change is exposing an, admittedly subtle, bug. That said, it is always possible there is a `clang` bug. Is `holopy` doing any CI on macOS?

```
======================================================================
ERROR: holopy.scattering.tests.test_multisphere.test_overlap
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/travis/miniconda/conda-bld/holopy_1498481066320/_t_env/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/travis/miniconda/conda-bld/holopy_1498481066320/_t_env/lib/python3.6/site-packages/holopy/scattering/tests/test_multisphere.py", line 170, in test_overlap
    holo = calc_holo(schema, sc, index, wavelen, xpolarization)
  File "/Users/travis/miniconda/conda-bld/holopy_1498481066320/_t_env/lib/python3.6/site-packages/holopy/scattering/calculations.py", line 134, in calc_holo
    scat = theory._calc_field(scatterer.guess(), uschema)
  File "/Users/travis/miniconda/conda-bld/holopy_1498481066320/_t_env/lib/python3.6/site-packages/holopy/scattering/theory/scatteringtheory.py", line 99, in _calc_field
    field = get_field(scatterer)
  File "/Users/travis/miniconda/conda-bld/holopy_1498481066320/_t_env/lib/python3.6/site-packages/holopy/scattering/theory/scatteringtheory.py", line 81, in get_field
    field = np.vstack(self._raw_fields(stack_spherical(positions), s, medium_wavevec=wavevec(schema), medium_index=schema.medium_index, illum_polarization=schema.illum_polarization)).T
  File "/Users/travis/miniconda/conda-bld/holopy_1498481066320/_t_env/lib/python3.6/site-packages/holopy/scattering/theory/multisphere.py", line 227, in _raw_fields
    raise MultisphereFailure()
holopy.scattering.errors.MultisphereFailure: Multisphere calculations failed to converge, or returned NaN. this probably means your scatterer is unphysical.
----------------------------------------------------------------------
```

File: recipes/holopy/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    - numpy x.x
+    - scipy
+    - pyyaml
+    - pillow
+    - h5py
+    - emcee
+    - matplotlib
+    - xarray >=0.9
+    - h5netcdf
+    - libgcc
```
Comment (jakirkham) [inline recipes/holopy/meta.yaml]: Would try changing this to `libgfortran`.

**Comment by barkls** [comment]
Thanks for the look. We are seeing different behaviour in CircleCI vs Travis so I guess there are two problems here.

On CircleCI, the code fails to compile, possibly due to toolchain (since everything runs fine without it). 

On Travis, one of our tests fails with np112py36, although it ran with both np112py35 and np113py35. I will look into this.

**Comment by barkls** [comment]
We do not have any CI testing on macs of our own, but we have built holopy and run tests on physical mac machines.

**Comment by barkls** [comment]
Hi @jakirkham,

 I've temporarily disabled the offending test on Travis, since the failure is not reproducible with identical books on my own branch, and the code it tests is soon-to-be-deprecated anyway. 

The CircleCI compile issue remains, and that is the one I attribute to ``toolchain`` since the package compiled before adding it as a dependency. 

Any tips?

**Comment by barkls** [comment]
Any help in resolving the CircleCI issue here @jakirkham? It compiled properly before adding the ``toolchain`` dependency, but hasn't worked since.

**Comment by isuruf** [comment]
Is this using `numpy.distutils`? If so, you have 2 options,

1. Use `toolchain3` instead of `toolchain`. It is still experimental though. 
2. `unset LDFLAGS` for Linux

**Comment by barkls** [comment]
yes, we are using ``numpy.distutils``. I will try those suggestions. Thanks!

File: recipes/holopy/build.sh (hunk 5 above)
Comment (isuruf) [inline recipes/holopy/build.sh]: I don't think you'd need any of these with `toolchain3`

**Comment by barkls** [comment]
@isuruf toolchain3 seems to work. For reference, I had to switch runtime dependency from libgfortran to libgcc in order to find libquadmath.o

@jakirkham Build is now passing on all platforms. What's the next step?

**Comment by jakirkham** [comment]
> For reference, I had to switch runtime dependency from libgfortran to libgcc in order to find libquadmath.o

On Mac, `libgfortran` pulls in `libquadmath`. So `libgfortran` should be all that is needed on Mac. As for Linux having `libgcc` outright is fine. Please update the recipe accordingly.

File: recipes/holopy/build.sh [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+if [ $(uname) == Darwin ]; then
+    export CFLAGS="${CFLAGS} -arch x86_64"
+    export FFLAGS="${FFLAGS} -static -ff2c -arch x86_64"
+    export LDFLAGS="${LDFLAGS} -Wall -undefined dynamic_lookup -bundle -arch x86_64"
+    export LDFLAGS="${LDFLAGS} -Wl,-search_paths_first -L$(pwd) $LDFLAGS"
+fi
```
Comment (isuruf) [inline recipes/holopy/build.sh]: Can you try removing all of the stuff above? I don't think any of it is needed with `toolchain3`.

**Comment by barkls** [comment]
Thanks for the review @isuruf. You were correct that I was able to remove the need for a ``build.sh`` file. Any other fixes?

File: recipes/holopy/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    It has tools for managing data and metadata, doing reconstructions, computing light scattering
+    from micron scale particles or clusters of such particles, and fitting scattering models to
+    digital holograms to measure particle parameters. HoloPy is free and open source software (GPL v3).
+  doc_url: http://holopy.readthedocs.io/
+  dev_url: https://github.com/manoharan-lab/holopy
+
+extra:
+  recipe-maintainers:
+    - barkls
+    - vnmanoharan
```
Comment (jakirkham) [inline recipes/holopy/meta.yaml]: I think we confirmed this before (maybe in the old PR), but just to check. Are you ok being a maintainer here, @vnmanoharan?

Comment (vnmanoharan) [inline recipes/holopy/meta.yaml]: yes, I'm OK with being a maintainer.

**Comment by barkls** [comment]
@jakirkham is there anything else we need to do on our end to get the feedstock up and running?

**Comment by jakirkham** [comment]
Nope. Just been swamped with other things.

**Comment by jakirkham** [comment]
Thanks @barkls for the contribution and @isuruf for helping with the review.

**Comment by barkls** [comment]
Thank you for all the help @jakirkham!

**Pull request merged**